web3.js/src/connection.ts

```diff
@@ -4281,75 +4232,209 @@ export class Connection {
       return;
     }
 
-    for (let id of accountKeys) {
-      const sub = this._accountChangeSubscriptions[id];
-      this._subscribe(
-        sub,
-        'accountSubscribe',
-        this._buildArgs([sub.publicKey], sub.commitment, 'base64'),
-      );
-    }
-
-    for (let id of programKeys) {
-      const sub = this._programAccountChangeSubscriptions[id];
-      this._subscribe(
-        sub,
-        'programSubscribe',
-        this._buildArgs([sub.programId], sub.commitment, 'base64', {
-          filters: sub.filters,
-        }),
-      );
-    }
-
-    for (let id of slotKeys) {
-      const sub = this._slotSubscriptions[id];
-      this._subscribe(sub, 'slotSubscribe', []);
-    }
-
-    for (let id of slotUpdateKeys) {
-      const sub = this._slotUpdateSubscriptions[id];
-      this._subscribe(sub, 'slotsUpdatesSubscribe', []);
-    }
-
-    for (let id of signatureKeys) {
-      const sub = this._signatureSubscriptions[id];
-      const args: any[] = [sub.signature];
-      if (sub.options) args.push(sub.options);
-      this._subscribe(sub, 'signatureSubscribe', args);
-    }
+    await Promise.all(
+      Object.entries(
+        this._subscriptionsByHash as Record<
+          SubscriptionConfigHash,
+          Subscription
+        >,
+      ).map(async ([hash, subscription]) => {
+        switch (subscription.state) {
+          case 'pending':
+          case 'unsubscribed':
+            if (subscription.callbacks.size === 0) {
+              /**
+               * You can end up here when:
+               *
+               * - a subscription has recently unsubscribed
+               *   without having new callbacks added to it
+               *   while the unsubscribe was in flight, or
+               * - when a pending subscription has its
+               *   listeners removed before a request was
+               *   sent to the server.
+               *
+               * Being that nobody is interested in this
+               * subscription any longer, delete it.
+               */
+              delete this._subscriptionsByHash[hash];
+              if (subscription.state === 'unsubscribed') {
+                delete this._subscriptionCallbacksByServerSubscriptionId[
+                  subscription.serverSubscriptionId
+                ];
+              }
+              await this._updateSubscriptions();
+              return;
+            }
+            await (async () => {
+              const {args, method} = subscription;
+              try {
+                this._subscriptionsByHash[hash] = {
+                  ...subscription,
+                  state: 'subscribing',
+                };
```

Comment:
By creating a new object here, I believe we could resubscribe multiple times to the same subscription:
```
1A (no callbacks; gets deleted then recurses)
 L 2B (pending, creates new object with subscribing state)
2A (old object is still pending, transitions to subscribing again)
```
I would have expected you to set the state directly on the existing object so that it serves as a lock to prevent multiple subscribes / unsubscribes.

Reply:
![](https://media1.giphy.com/media/26uf2NLT7fDRPwl7G/200.gif?cid=5a38a5a2y7alhu3mxhwar8nalpwd5bqzeuue5hmmim08v2r3&amp;rid=200.gif&amp;ct=g)

OK. `Object.entries`… not the right choice here. Switching now to a strategy where what we iterate over are the hashes, then do the lookup as late as possible.